### PR TITLE
zsh: make zpty module work

### PR DESCRIPTION
**Pull request by SquallATF**
make zsh/zpty work with msys2, and can pass most mafredri/zsh-async#26 test.

```c
    master = movefd(master);
    if (master == -1) {
	zerrnam(nam, "cannot duplicate fd %d: %e", master, errno);
	scriptname = oscriptname;
	ineval = oineval;
	return 1;
    }
```

Since cygwin uses two threads to process the pseudoterminal master device, zpty calls the `movefd` method to move the file description of the master device, and cygwin disposed the original `fhandler_pty_master` object, which causes the process thread to run incorrectly. here simply disable the zpty move the master device file descriptor untill cygwin upstream fix this issue.

-  zpty error strace

```
   30 27794711 [main] zsh 1918 fhandler_pty_master::cleanup: /dev/ptmx closing master, usecount 0
   43 27794754 [main] zsh 1918 fhandler_pty_master::close: closing from_master(0x310)/from_master_cyg(0x310)/to_master(0x264)/to_master_cyg(0x31C) since we own them(0)
   38 27794792 [main] zsh 1918 fhandler_pty_master::close: (2420): pty output_mutex (0x1F8): waiting -1 ms
   37 27794829 [main] zsh 1918 fhandler_pty_master::close: (2420): pty output_mutex: acquired
   71 27794900 [ptym] zsh 1918 fhandler_pty_master::pty_master_thread: ReadFile, Win32 error 6
   34 27794934 [ptym] zsh 1918 fhandler_pty_master::pty_master_thread: Reply: from 0x0, to 0x0, error 6
   37 27794971 [ptym] zsh 1918 fhandler_pty_master::pty_master_thread: WriteFile, Win32 error 6
   37 27795008 [ptym] zsh 1918 fhandler_pty_master::pty_master_thread: DisconnectNamedPipe, Win32 error 6
   39 27795047 [ptym] zsh 1918 fhandler_pty_master::pty_master_thread: Leaving
```

-  zsh-async test log

```
=== RUN   test__async_job_multiple_commands
--- PASS: test__async_job_multiple_commands (0.08s)
=== RUN   test__async_job_print_hi
--- PASS: test__async_job_print_hi (0.07s)
=== RUN   test__async_job_stderr
--- PASS: test__async_job_stderr (0.07s)
=== RUN   test__async_job_wait_for_token
--- PASS: test__async_job_wait_for_token (0.19s)
=== RUN   test_all_options
--- FAIL: test_all_options (15.05s)
        ./async_test.zsh:517: timed out
=== RUN   test_async_flush_jobs
--- PASS: test_async_flush_jobs (0.63s)
=== RUN   test_async_job_error_and_nonzero_exit
--- PASS: test_async_job_error_and_nonzero_exit (0.24s)
=== RUN   test_async_job_git_status
--- PASS: test_async_job_git_status (0.43s)
=== RUN   test_async_job_keeps_nulls
--- PASS: test_async_job_keeps_nulls (0.21s)
=== RUN   test_async_job_multiple_arguments_and_spaces
--- PASS: test_async_job_multiple_arguments_and_spaces (0.27s)
=== RUN   test_async_job_multiple_commands_in_multiline_string
--- PASS: test_async_job_multiple_commands_in_multiline_string (0.22s)
=== RUN   test_async_job_print_matches_input_exactly
--- PASS: test_async_job_print_matches_input_exactly (0.21s)
=== RUN   test_async_job_unique_worker
--- PASS: test_async_job_unique_worker (0.49s)
=== RUN   test_async_job_with_rc_expand_param
--- PASS: test_async_job_with_rc_expand_param (0.19s)
=== RUN   test_async_process_results
--- PASS: test_async_process_results (0.25s)
=== RUN   test_async_process_results_stress
--- FAIL: test_async_process_results_stress (9.13s)
        ./async_test.zsh:182: timed out after 5s
        ./async_test.zsh:183: wanted 20 results, got 14
=== RUN   test_async_start_stop_worker
--- PASS: test_async_start_stop_worker (0.06s)
=== RUN   test_async_worker_notify_sigwinch
--- PASS: test_async_worker_notify_sigwinch (2.44s)
=== RUN   test_async_worker_survives_termination_of_other_worker
--- PASS: test_async_worker_survives_termination_of_other_worker (0.15s)
=== RUN   test_async_worker_update_pwd
--- PASS: test_async_worker_update_pwd (0.33s)
=== RUN   test_async_worker_update_pwd_and_env
--- PASS: test_async_worker_update_pwd_and_env (0.36s)
=== RUN   test_zle_watcher
--- SKIP: test_zle_watcher (0.01s)
        ./async_test.zsh:596: Test is not reliable on zsh 5.0.X
FAIL
exit code 1
FAIL    ./async_test.zsh        31.139s
```

### CODE REVIEW

**Comment by Biswa96** [comment]
Run `updpkgsums` which will update the checksums automatically, then force push the changes here.

File: zsh/fix-zpty-module.patch [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+diff --git a/configure.ac b/configure.ac
+index bd1b40225..3b0cc1f53 100644
+--- a/configure.ac
++++ b/configure.ac
+@@ -2504,7 +2504,7 @@ if test x$ac_cv_have_dev_ptmx = xyes -o x$ac_cv_func_posix_openpt = xyes && \
+    test x$ac_cv_func_ptsname = xyes; then
+    AC_CACHE_CHECK([if /dev/ptmx is usable],
+    ac_cv_use_dev_ptmx,
+-   [AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#ifdef __linux
++   [AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#if definde(__linux) || defined(__CYGWIN__)
```
Comment (lazka) [inline zsh/fix-zpty-module.patch:30]: typo

**Comment by lazka** [comment]
Is there any upstream issue for this (zsh or cygwin)?

**Comment by SquallATF** [comment]
There is a cygwin upstream issue but patched zpty to avoid this issue.
zpty invoke `movefd` will dup and close original fd of `/dev/ptmx` and cause the `fhandler_pty_master` object to be recycled,  which will damage `pty_master_thread` and `pty_master_fwd_thread` threads on that object.

File: zsh/PKGBUILD [hunk 2]
```diff
@@ -43,6 +45,7 @@ prepare() {
   patch -p1 -i "${srcdir}/msysize.patch"
   patch -p1 -i "${srcdir}/zsh-5.0.6-1.patch"
   patch -p1 -i "${srcdir}/msys-symlink-hack.patch"
+  patch -p1 -i "${srcdir}/fix-zpty-module.patch"
```
Comment (lazka) [inline zsh/PKGBUILD:48]: Can you please add a comment above this with a link to the upstream issue. So we know when we can drop the patch.

Comment (SquallATF) [reply zsh/PKGBUILD:48]: ok, but `configure.ac` change is needed for cygwin/msys2 to use ptmx device, if zsh upstream not fix.

**Pull request merged**